File: euclidean_primer/euclidean_primer__rsa_solve.py
# 実行時間 10ms
def extGcd(a, b):
    if b != 0:
        c, y, x = extGcd(b, a % b)

        y -= a // b * x
        return c, x, y
    return a, 1, 0

p, q, e, f = list(map(int, input().split()))
n = p * q
nd = (p - 1) * (q - 1)
x = extGcd(e, nd)[1]

print(chr(pow(f, (nd + x) % nd, n)))

File: euclidean_primer/euclidean_primer__rsa_solve.cs
using System;
using static System.Console;

internal class Program
{
    static void Main()
    {
        long[] pqeE = Array.ConvertAll( ReadLine().Split(), long.Parse );
        long n = pqeE[0] * pqeE[1];
        long nd = ( pqeE[0] - 1 ) * ( pqeE[1] - 1 );
        long x = 1;
        long y = 1;

        ExtGcd( pqeE[2], nd, ref x, ref y );
        WriteLine( Convert.ToChar( ModPow( pqeE[3], ( nd + x ) % nd, n ) ) );
        return;
    }

    static long ExtGcd( long a, long b, ref long x, ref long y )
    {
        long c = a;

        if( b != 0 ) {
            c = ExtGcd( b, a % b, ref y, ref x );
            y -= a / b * x;
        } else {
            x = 1;
            y = 0;
        }
        return c;
    }

    static ulong ModPow( long a, long b, long m )
    {
        ulong la = (ulong)a;
        ulong lb = (ulong)b;
        ulong lm = (ulong)m;
        ulong c = 1;

        while( 0 < lb ) {
            if( ( lb & 1 ) == 1 ) {
                c = c * la % lm;
            }
            la = la * la % lm;
            lb >>= 1;
        }
        return c;
    }
}
